pkg/controller/v1alpha1/localmodel/controller.go

```diff
@@ -0,0 +1,658 @@
+		},
+		CreateFunc: func(e event.CreateEvent) bool {
+			if _, ok := e.Object.GetLabels()[constants.LocalModelLabel]; !ok {
+				return false
+			}
+			return true
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			if _, ok := e.Object.GetLabels()[constants.LocalModelLabel]; !ok {
+				return false
+			}
+			return true
+		},
+	}
+
+	nodePredicates := predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			// Only reconciles the local model crs when the node becomes ready from not ready
+			// Todo: add tests
+			oldNode := e.ObjectNew.(*v1.Node)
+			newNode := e.ObjectNew.(*v1.Node)
+			return !isNodeReady(*oldNode) && isNodeReady(*newNode)
+		},
+		CreateFunc: func(e event.CreateEvent) bool {
+			// Do nothing here, generates local model cr reconcile requests in nodeFunc
+			return true
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return false
+		},
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&v1alpha1api.ClusterLocalModel{}).
+		// Ownes Jobs, PersistentVolumes and PersistentVolumeClaims that is created by this local model controller
+		Owns(&batchv1.Job{}).
+		Owns(&v1.PersistentVolume{}).
+		Owns(&v1.PersistentVolumeClaim{}).
+		// Creates or deletes pv/pvcs when isvcs got created or deleted
+		Watches(&v1beta1.InferenceService{}, handler.EnqueueRequestsFromMapFunc(c.isvcFunc), builder.WithPredicates(isvcPredicates)).
+		// Downloads models to new nodes
+		Watches(&v1.Node{}, handler.EnqueueRequestsFromMapFunc(c.nodeFunc), builder.WithPredicates(nodePredicates)).
+		Complete(c)
+}
+
+func (c *LocalModelReconciler) launchDeletionJob(jobName string, namespace string, localModel *v1alpha1api.ClusterLocalModel, storageUri string, claimName string, node string) (*batchv1.Job, error) {
+	container, err := c.getContainerSpecForStorageUri(storageUri)
+	if err != nil {
+		return nil, err
+	}
+	container.Command = []string{"/bin/sh", "-c", "rm -rf /mnt/models/*"}
+	container.Args = nil
+	return c.launchJob(jobName, *container, namespace, localModel, storageUri, claimName, node)
+}
+
+// Launches a job if not exist, or return the existing job
+func (c *LocalModelReconciler) launchDownloadJob(jobName string, namespace string, localModel *v1alpha1api.ClusterLocalModel, storageUri string, claimName string, node string) (*batchv1.Job, error) {
+	container, err := c.getContainerSpecForStorageUri(storageUri)
+	if err != nil {
+		return nil, err
+	}
+
+	return c.launchJob(jobName, *container, namespace, localModel, storageUri, claimName, node)
+}
+
+// Launches a job if not exist, or return the existing job
+func (c *LocalModelReconciler) launchJob(jobName string, container v1.Container, namespace string, localModel *v1alpha1api.ClusterLocalModel, storageUri string, claimName string, node string) (*batchv1.Job, error) {
+	jobs := c.Clientset.BatchV1().Jobs(namespace)
+
+	job, err := jobs.Get(context.TODO(), jobName, metav1.GetOptions{})
```

Comment:
I found an issue that if the job was failed to launch then we are not able to rerun it because it checks the job name which is fixed. If the job is not completed then we should recreate the job.

Reply:
Should be fixed.